Generate CGColor serialization

beidson opened the pull request:
#### 2491896493663e307f90dc6be53672e9ca59c4ef
<pre>
Generate CGColor serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=263932">https://bugs.webkit.org/show_bug.cgi?id=263932</a>
rdar://117709336

Reviewed by David Kilzer.

WebCore::Color is already an appropriate wrapper class for CGColorRef.

But to use it, I needed to change generate-serializers.py to allow for:
- Fully custom &quot;toCF&quot; function
- Custom &quot;fromCF&quot; function
- Add an arbitrary forward declaration to the generated header.

With those changes, the actual code changes to WebKit are pure deletion.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(ConditionalForwardDeclaration):
(ConditionalForwardDeclaration.__init__):
(ConditionalForwardDeclaration.__lt__):
(ConditionalForwardDeclaration.__lt__.condition_str):
(ConditionalForwardDeclaration.__eq__):
(ConditionalForwardDeclaration.__hash__):
(generate_header):
(encode_cf_type):
(decode_cf_type):
(parse_serialized_types):
(main):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;CFBarRef&gt;::encode):
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFBarRef&gt;&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder&lt;CGColorRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;CGColorRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/269987@main">https://commits.webkit.org/269987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9a5b44fc965f23d9c64be34b2ec878b52a1f951

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24187 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25271 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26321 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24457 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24659 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24431 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20904 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26910 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/24359 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21823 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22048 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22118 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/19172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5794 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/1911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1870 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
```diff
@@ -73,6 +73,11 @@ namespace JSC { enum class Incredible; }
 namespace Testing { enum class StorageSize : uint8_t; }
 namespace WebCore { class ScrollingStateFrameHostingNode; }
 namespace WebCore { class ScrollingStateFrameHostingNodeWithStuffAfterTuple; }
+#if USE(CFBAR)
+#endif
```
Can we avoid outputting this empty `#if USE(CFBAR)/#endif` code?

Would prefer to fix this before landing, but not strictly required to do so.

beidson replied:
What this patch doesn't make obvious is that this issue is long standing
If you looked at your <builddir>/DerivedSources/WebKit/GeneratedSerializers files you'd see tons of these repeated empty conditionals.

I agreed cleaning them would be great, but should be a standalone hack to the serializer script.

ddkilzer commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCSerialization.mm:
```diff
@@ -74,7 +74,9 @@ CFTypeRef valueAsCFType() const
     typedef std::variant<
         RetainPtr<CFStringRef>,
         RetainPtr<CFURLRef>,
-        RetainPtr<CFDataRef>
+        RetainPtr<CFDataRef>,
+        RetainPtr<CFBooleanRef>,
+        RetainPtr<CGColorRef>
```
Nit: Would be nice if these were sorted.  Not required to land PR.

beidson replied:
(As you guessed) we're just adding them chronologically as we go.
I'll land this one as-is, and we can start sorting next time.

ddkilzer reviewed (approved):
r=me

beidson committed:
Generate CGColor serialization
https://bugs.webkit.org/show_bug.cgi?id=263932
rdar://117709336

Reviewed by David Kilzer.

WebCore::Color is already an appropriate wrapper class for CGColorRef.

But to use it, I needed to change generate-serializers.py to allow for:
- Fully custom "toCF" function
- Custom "fromCF" function
- Add an arbitrary forward declaration to the generated header.

With those changes, the actual code changes to WebKit are pure deletion.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(ConditionalForwardDeclaration):
(ConditionalForwardDeclaration.__init__):
(ConditionalForwardDeclaration.__lt__):
(ConditionalForwardDeclaration.__lt__.condition_str):
(ConditionalForwardDeclaration.__eq__):
(ConditionalForwardDeclaration.__hash__):
(generate_header):
(encode_cf_type):
(decode_cf_type):
(parse_serialized_types):
(main):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<CFBarRef>::encode):
(IPC::ArgumentCoder<RetainPtr<CFBarRef>>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<CGColorRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<CGColorRef>>::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCSerialization.mm:
(TEST):

Canonical link: https://commits.webkit.org/269987@main
